src/com/magento/idea/magento2plugin/actions/generation/dialog/NewSetupDataPatchDialog.java

```diff
@@ -0,0 +1,223 @@
+    }
+
+    /**
+     * Open dialog.
+     *
+     * @param project Project
+     * @param directory PsiDirectory
+     */
+    public static void open(
+            final Project project,
+            final PsiDirectory directory,
+            final String modulePackage,
+            final String moduleName
+    ) {
+        final NewSetupDataPatchDialog dialog = new NewSetupDataPatchDialog(
+                project,
+                directory,
+                modulePackage,
+                moduleName
+        );
+        dialog.pack();
+        dialog.centerDialog(dialog);
+        dialog.setVisible(true);
+    }
+
+    protected void onOK() {
+        if (validateFields()) {
+            generateFile();
+            exit();
+        }
+    }
+
+    private void generateFile() {
+        final ModuleSetupDataPatchGenerator generator = new ModuleSetupDataPatchGenerator(
+                new ModuleSetupDataPatchData(
+                        modulePackage,
+                        moduleName,
+                        createDirectory(baseDir),
+                        getClassName()
+                ),
+                project
+        );
+
+        generator.generate(NewSetupDataPatchAction.ACTION_NAME, true);
+    }
+
+    public String getClassName() {
+        return className.getText().trim();
+    }
+
+    private boolean validateFields() {
+        final PsiDirectory targetDirectory = getDirectory(baseDir);
+
+        if (NewSetupDataPatchAction.DATA_DIRECTORY.equals(targetDirectory.getName())) {
+            final PsiFile[] files = targetDirectory.getFiles();
+            for (final PsiFile file : files) {
+                final String className = ModuleSetupDataPatchFile.resolveClassNameFromInput(
+                        getClassName()
+                );
+
+                if (file.getName().equals(className + ".php")) {
+                    showErrorMessage(
+                            fieldsValidationsList.get(0).getField(),
+                            "Class name `" + className + "` already exist."
+                    );
+
+                    return false;
+                }
+            }
+        }
+        return validateFormFields();
+    }
+
+    private PsiDirectory getDirectory(final PsiDirectory targetDirectory) {
+        if (NewSetupDataPatchAction.ROOT_DIRECTORY.equals(baseDir.getName())) {
+            final PsiDirectory subDirectoryPatch = baseDir.findSubdirectory(
+                    NewSetupDataPatchAction.PATCH_DIRECTORY
+            );
+
+            if (subDirectoryPatch != null) {
+                return subDirectoryPatch.findSubdirectory(NewSetupDataPatchAction.DATA_DIRECTORY);
+            }
+        }
+        if (NewSetupDataPatchAction.PATCH_DIRECTORY.equals(baseDir.getName())) {
+            return baseDir.findSubdirectory(NewSetupDataPatchAction.DATA_DIRECTORY);
+        }
+
+        return targetDirectory;
+    }
+
+    private PsiDirectory createDirectory(final PsiDirectory targetDirectory) {
```

Comment:
Please simplify this logic. Try to use `com.magento.idea.magento2plugin.actions.generation.generator.util.DirectoryGenerator#findOrCreateSubdirectories`

Reply:
Done.